File: web/src/lib/media/chunk.ts
export const CHUNK_SIZE = 64 * 1024; // 64KB (Safari DataChannel 안전 범위)

/**
 * 파일을 64KB 청크로 분할
 */
export function splitIntoChunks(data: Uint8Array): Uint8Array[] {
  const chunks: Uint8Array[] = [];
  for (let offset = 0; offset < data.length; offset += CHUNK_SIZE) {
    chunks.push(data.slice(offset, offset + CHUNK_SIZE));
  }
  return chunks;
}

/**
 * 청크 배열을 하나의 Uint8Array로 재조립
 */
export function reassembleChunks(
  chunks: Map<number, Uint8Array>,
  totalChunks: number
): Uint8Array {
  let totalLength = 0;
  for (let i = 0; i < totalChunks; i++) {
    const chunk = chunks.get(i);
    if (!chunk) throw new Error(`Missing chunk ${i}`);
    totalLength += chunk.length;
  }

  const result = new Uint8Array(totalLength);
  let offset = 0;
  for (let i = 0; i < totalChunks; i++) {
    const chunk = chunks.get(i)!;
    result.set(chunk, offset);
    offset += chunk.length;
  }

  return result;
}

/**
 * SHA-256 해시 계산 (무결성 검증)
 */
export async function computeChecksum(data: Uint8Array): Promise<string> {
  // @ts-expect-error -- TS strict ArrayBuffer typing vs runtime Uint8Array acceptance
  const hash = await crypto.subtle.digest('SHA-256', data);
  return arrayBufferToHex(hash);
}

/**
 * checksum 검증
 */
export async function verifyChecksum(
  data: Uint8Array,
  expectedChecksum: string
): Promise<boolean> {
  const actual = await computeChecksum(data);
  return actual === expectedChecksum;
}

function arrayBufferToHex(buffer: ArrayBuffer): string {
  return Array.from(new Uint8Array(buffer))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

File: mobile/lib/core/media/chunk.dart
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

import '../constants/app_constants.dart';

/// 데이터를 64KB 청크로 분할
List<Uint8List> splitIntoChunks(Uint8List data) {
  final chunks = <Uint8List>[];
  for (var offset = 0; offset < data.length; offset += AppConstants.chunkSize) {
    final end = (offset + AppConstants.chunkSize).clamp(0, data.length);
    chunks.add(Uint8List.fromList(data.sublist(offset, end)));
  }
  return chunks;
}

/// 청크 재조립
Uint8List reassembleChunks(Map<int, Uint8List> chunks, int totalChunks) {
  var totalLength = 0;
  for (var i = 0; i < totalChunks; i++) {
    final chunk = chunks[i];
    if (chunk == null) throw StateError('Missing chunk $i');
    totalLength += chunk.length;
  }

  final result = Uint8List(totalLength);
  var offset = 0;
  for (var i = 0; i < totalChunks; i++) {
    final chunk = chunks[i]!;
    result.setRange(offset, offset + chunk.length, chunk);
    offset += chunk.length;
  }

  return result;
}

/// SHA-256 체크섬 (hex string)
String computeChecksum(Uint8List data) {
  return sha256.convert(data).toString();
}

/// 체크섬 검증
bool verifyChecksum(Uint8List data, String expectedChecksum) {
  return computeChecksum(data) == expectedChecksum;
}
